src/SelfGravity/Poi_UserWorkBeforePoisson_GREP.cpp

```diff
@@ -176,81 +242,162 @@ void Mis_UserWorkBeforeNextSubstep_GREP( const int lv, const double TimeNew, con
 // Description :  Update the spherical-averaged profiles before Poisson and Gravity solvers,
-//                and compute the GR effective potential.
+//                and compute the effective GR potential.
 //
 // Note        :  1. Invoked by Poi_UserWorkBeforePoisson_GREP()
-//                2. The contribution from     leaf patches on level = lv (<= lv) is stored QUANT[    lv]
-//                                         non-leaf patches on level = lv         is stored QUANT[NLEVEL]
-//                3. The GR effective potential is stored at Phi_eff[lv]
+//                2. The contributions from     leaf patches on level=lv are stored in GREP_*Ave[    lv]
+//                                          non-leaf patches                           GREP_*Ave[NLEVEL]
+//                3. The effective GR potential is stored at GREP_EffPot[lv]
 //
 // Parameter   :  Time : Target physical time
 //                lv   : Target refinement level
 //-------------------------------------------------------------------------------------------------------
 void Poi_Prepare_GREP( const double Time, const int lv )
 {
 
-// compare the input Time with stored time to choose the suitable SaveSg
+// compare the input time with the stored time to determine the appropriate sandglass
    int Sg;
 
-   if      (  Mis_CompareRealValue( Time, GREPSgTime[lv][0], NULL, false )  )   Sg = 0;
-   else if (  Mis_CompareRealValue( Time, GREPSgTime[lv][1], NULL, false )  )   Sg = 1;
-   else                                                                         Sg = 1 - GREPSg[lv];
+   if      (  Mis_CompareRealValue( Time, GREP_SgTime[lv][0], NULL, false )  )   Sg = 0;
+   else if (  Mis_CompareRealValue( Time, GREP_SgTime[lv][1], NULL, false )  )   Sg = 1;
+   else                                                                          Sg = 1 - GREP_Sg[lv];
 
 
-// update and combine the spherical-averaged profiles
-   long       TVar         [] = {               _DENS,             _VELR,               _PRES,               _EINT };
-   Profile_t *Prof_Leaf    [] = { DensAve[    lv][Sg], VrAve[    lv][Sg], PresAve[    lv][Sg], EngyAve[    lv][Sg] };
-   Profile_t *Prof_NonLeaf [] = { DensAve[NLEVEL][Sg], VrAve[NLEVEL][Sg], PresAve[NLEVEL][Sg], EngyAve[NLEVEL][Sg] };
+// update the spherical-averaged profiles contributed from non-leaf and leaf patches
+   GREP_Compute_Profile( lv, Sg, PATCH_LEAF    );
+   GREP_Compute_Profile( lv, Sg, PATCH_NONLEAF );
 
 
-   if ( false ) {}
-   /*
-   if ( GREP_ALGO_SWITCH )
-   {
-//    contributions from the leaf patches on level <= lv and the non-leaf patches on level = lv
-      Aux_ComputeProfile( Prof_NonLeaf, GREP_Prof_Center, GREP_Prof_MaxRadius, GREP_Prof_MinBinSize,
-                          GREP_LOGBIN, GREP_LOGBINRATIO, true,  TVar, 4,  0, lv, PATCH_LEAF_PLUS_MAXNONLEAF, Time );
-   }
-   */
+// combine the spherical-averaged profiles
+   GREP_Combine_Profile( GREP_DensAve, lv, Sg, Time, true );
```

Comment:
Define `const bool RemoveEmpty_Yes = true;` for clarity.

Reply:
Done